llvm/lib/CASObjectFormats/NestedV1.cpp

```diff
@@ -2129,5 +2129,6 @@ Error LinkGraphBuilder::defineSymbol(SymbolToDefine Symbol) {
 
   G.redefineSymbol(*Symbol.S, **Block, Symbol.Ref.getOffset(),
                    /*Size=*/0, Linkage, Scope, IsLive);
+  Symbol.S->setAutoHide(MergeByContent);
```

Comment:
This is insufficient; `AutoHide` will be set too often once we set `MergeByContent` in all the relevant cases (right now, it's only set for AutoHide because the Mach-O has lost information necessary to set it elsewhere; or maybe the code is setting it in extra places (unsoundly) already, not sure). I'm not sure what `lld` will do with the "auto-hide" bit when we start putting `MergeByContent` on more things.

LLVM writes AutoHide into Mach-O when both:
- linkage is `linkonce_odr`
- either
    - `unnamed_addr`
    - both
        - `local_unnamed_addr`
        - runtime constant (not a global variable, or a constant global variable)

MergeByContent is valid when either:
- `unnamed_addr`
- both
    - `local_unnamed_addr`
    - runtime constant (not a global variable, or a constant global variable)

(i.e., the linkage is not relevant).

The absolutely safe check to recover AutoHide is all three of:
- `DeadStrip == Compile`
    - not in `llvm.used`
    - one of `linkonce*`, `internal`, or `private`
- `Scope != Local`
    - must be exported (not `internal` or `private`)
- `Merge == ByNameOrContent`
    -  ByName means `weak*` or `linkonce*`
    -  ByContent indicates at least `local_unnamed_addr`

Pedantically, requiring `DeadStrip=Compile` is too conservative, since it'll miss setting the autohide bit on `linkonce*` functions that are in `llvm.used`. But I doubt this matters in practice.

Reply:
How about I add the `isAutoHide` property as an extra bit in `SymbolRef::Flags`? I think it's better to have this symbol property passing through decoupled from what effects it's going to have or not.

Reply:
I'm not sure I agree.

AutoHide passes weirdly specific (and lossy) information from LLVM to the linker: hey ld64, here's a place where your `-dedup` algorithm is correct, but I'm only telling you about this one because (a) Mach-O doesn't have enough bits to tell you about the others and (b) your algorithm is too slow to run in all the places it's safe because embedded fixups in Mach-O require you to do heroics to compare block content.

Instead, I think we should be storing something higher-fidelity, with independent axes separated, allowing the linker to receive all the information LLVM is able to provide and then do whatever it wants with it.

Adding an autohide bit here defeats that goal.

For now, I suggest:
- Add an `isAutoHide()` accessor on SymbolRef that returns `true` in the places I mention above, and have `lld` use that.
- Add FIXME in the doxygen comments for SymbolRef for passing more information through from LLVM.
- Add FIXME in lld for somehow taking advantage of MergeByContent (once it's set more often), rather than only in the specific case of AutoHide.

> Pedantically, requiring `DeadStrip=Compile` is too conservative, since it'll miss setting the autohide bit on `linkonce*` functions that are in `llvm.used`. But I doubt this matters in practice.

As a follow-up, the reason it doesn't matter in practice is that the linker isn't allowed to drop the symbol from the symbol table anyway if it's marked `used`. I could construct a case where linker output would change but it's not really going to be worse.

Reply:
> I could construct a case where linker output would change but it's not really going to be worse.

(and IMO, would be better fixed by properly supporting `MergeByContent` rather than strict adherence to AutoHide)

Reply:
> Add an isAutoHide() accessor on SymbolRef that returns true in the places I mention above, and have lld use that.

To make sure I understand, is the idea that "LLVM writing AutoHide into Mach-O" is not happening in all the symbols that could be considered `AutoHide`, and the CAS roundtrip will set `AutoHide` in more places?

> Add FIXME in the doxygen comments for SymbolRef for passing more information through from LLVM.

Could you provide more details, do you mean passing through the info from the conditions that you outlined for returning `isAutoHide()`, or do you have something other in mind?

> Add FIXME in lld for somehow taking advantage of MergeByContent (once it's set more often), rather than only in the specific case of AutoHide.

`MergeByContent` is already taken advantage of, meaning if symbols share the same `jitlink::Block` instance then the symbols in the binary share the same code block. I didn't specifically check for it in tests but I could enhance the tests for that.
Or do you have something else in mind?

Reply:
All done! I've also added a [FIXME](https://github.com/apple/llvm-project/pull/3557/files#diff-d43fd0dc0534175bec51f6a4945a5e3f868e106c0068639df12404fd3e6241acR2105) about the sharing of `jitlink::Block` pointers for merged symbols not being ideal without more information going to the client.